File: DemoEasypermissions/app/src/main/java/com/tilismtech/demoeasypermissions/MainActivity.java
package com.tilismtech.demoeasypermissions;

import android.Manifest;
import android.app.AlertDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import java.util.List;

import pub.devrel.easypermissions.EasyPermissions;

public class MainActivity extends AppCompatActivity implements EasyPermissions.PermissionCallbacks{

    private int REQUEST_CAMERA = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Button button = (Button)findViewById(R.id.button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                scanQr();
            }
        });
    }


    public void scanQr() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (!EasyPermissions.hasPermissions(this, Manifest.permission.CAMERA)) {
                EasyPermissions.requestPermissions(this, getString(R.string.camera_perm_for_qr), REQUEST_CAMERA, Manifest.permission.CAMERA);
            }
        }
        openQrScanActivity();
    }

    private void openQrScanActivity() {
        Toast.makeText(MainActivity.this,"Opening Camera Activity", Toast.LENGTH_SHORT).show();
    }


    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
        openQrScanActivity();
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        if (EasyPermissions.somePermissionPermanentlyDenied(this, perms)) {
            permissionDialog(getResources().getString(R.string.scan_qr_code), getResources().getString(R.string.camera_perm_for_qr));
        } else {
            Toast.makeText(MainActivity.this, R.string.camera_perm_for_qr, Toast.LENGTH_SHORT).show();
        }
    }

    private void permissionDialog(String title, String body) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(title);
        builder.setMessage(body);
        builder.setPositiveButton(R.string.ok, (dialog, which) -> {
            dialog.dismiss();
            Intent intent = new Intent();
            intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
            Uri uri = Uri.fromParts("package", getApplicationContext().getPackageName(), null);
            intent.setData(uri);
            startActivity(intent);
        });
        builder.setNegativeButton(R.string.cancel, (dialog, which) -> dialog.dismiss());
        builder.create().show();
    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CAMERA) {
            EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
        }
    }
}

File: DemoUnitTesting/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.R
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import pub.devrel.easypermissions.EasyPermissions
import pub.devrel.easypermissions.PermissionRequest


//import pub.devrel.easypermissions.EasyPermissions


class MainActivity : AppCompatActivity(), EasyPermissions.RationaleCallbacks,EasyPermissions.PermissionCallbacks {

    var rationalShown = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(com.example.myapplication.R.layout.activity_main)
//
        val perms =
            arrayOf(
                android.Manifest.permission.ACCESS_FINE_LOCATION,
                android.Manifest.permission.ACCESS_COARSE_LOCATION
            )
//
        if (EasyPermissions.hasPermissions(this, *perms)) {
            // Already have permission, do the thing
            // ...
        } else {

//            Request 1
            // Do not have permissions, request them now
//            EasyPermissions.requestPermissions(
//                this, "Required Permission",
//                0, *perms
//            )

//            Request 2
            EasyPermissions.requestPermissions(
                PermissionRequest.Builder(this, 0, *perms)
                    .setRationale("Location Request for konnect app")
                    .setPositiveButtonText("ok")
                    .setNegativeButtonText("Cancel")
                    .build()
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions!!, grantResults!!)
        // Forward results to EasyPermissions
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this)

        var deniedForever = false;

        for (permission in permissions) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
                //denied
                if(rationalShown)
                Log.d("Adnan", "Denied")
                else
                    Log.d("Adnan", "First time Denied")

            } else {
                if (ActivityCompat.checkSelfPermission(
                        this,
                        permission
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    //allowed
                    Log.d("Adnan", "Granted")
                } else {
                    //set to never ask again
                    Log.d("Adnan", "Set to never asked again")
                    deniedForever = true;
                }
            }
            if (deniedForever) {
                //show custom dialog and on click that dialog go to settings
                Log.d("Adnan", "Denied forever")
            }
        }


    }

    override fun onPermissionsGranted(requestCode: Int, perms: MutableList<String>) {
    }

    override fun onPermissionsDenied(requestCode: Int, perms: MutableList<String>) {
    }

    override fun onRationaleAccepted(requestCode: Int) {
        Log.d("Adnan", "Rational Accepted")
        rationalShown = true
    }

    override fun onRationaleDenied(requestCode: Int) {
        Log.d("Adnan", "Rational Denied")
    }
}